spinoso-string/src/enc/mod.rs

```diff
@@ -0,0 +1,462 @@
+mod ascii;
+mod binary;
+mod impls;
+mod utf8;
+
+use alloc::vec::Vec;
+use core::ops::Range;
+use core::slice::SliceIndex;
+
+use ascii::AsciiString;
+use binary::BinaryString;
+use utf8::Utf8String;
+
+use crate::codepoints::InvalidCodepointError;
+use crate::encoding::Encoding;
+use crate::iter::{Bytes, IntoIter, Iter, IterMut};
+use crate::ord::OrdError;
+
+#[derive(Clone)]
+pub enum EncodedString {
+    Ascii(AsciiString),
+    Binary(BinaryString),
+    Utf8(Utf8String),
+}
+
+impl Default for EncodedString {
+    fn default() -> Self {
+        Self::Utf8(Utf8String::new(Vec::new()))
+    }
+}
+
+// Constructors
+impl EncodedString {
+    #[inline]
+    #[must_use]
+    pub fn new(buf: Vec<u8>, encoding: Encoding) -> Self {
+        match encoding {
+            Encoding::Ascii => Self::Ascii(AsciiString::new(buf)),
+            Encoding::Binary => Self::Binary(BinaryString::new(buf)),
+            Encoding::Utf8 => Self::Utf8(Utf8String::new(buf)),
+        }
+    }
+}
+
+impl EncodedString {
```

Comment:
Note: The encoded types don't have a `encoding()` method. It felt like a bit of a circular reference. Theory is that those types would only really be used internally by String

Reply:
sounds good to me